go/extractor/cli/go-autobuilder/go-autobuilder.go

```diff
@@ -536,75 +554,120 @@ func installDependenciesAndBuild() {
+	for i, workspace := range workspaces {
+		goVersionInfo := workspace.RequiredGoVersion()
 
-	importpath := getImportPath()
-	needGopath := getNeedGopath(buildInfo, importpath)
+		fixGoVendorIssues(&workspace, goVersionInfo.Found)
 
-	inLGTM := os.Getenv("LGTM_SRC") != "" || os.Getenv("LGTM_INDEX_NEED_GOPATH") != ""
+		tryUpdateGoModAndGoSum(workspace)
 
-	if inLGTM && needGopath {
-		paths := moveToTemporaryGopath(srcdir, importpath)
+		// check whether an explicit dependency installation command was provided
+		inst := util.Getenv("CODEQL_EXTRACTOR_GO_BUILD_COMMAND", "LGTM_INDEX_BUILD_COMMAND")
+		shouldInstallDependencies := false
+		if inst == "" {
+			shouldInstallDependencies = buildWithoutCustomCommands(workspace.ModMode)
+		} else {
+			buildWithCustomCommands(inst)
+		}
 
-		// schedule restoring the contents of newdir to their original location after this function completes:
-		defer restoreRepoLayout(paths.newdir, paths.files, filepath.Base(paths.scratch), srcdir)
+		if workspace.ModMode == project.ModVendor {
+			// test if running `go` with -mod=vendor works, and if it doesn't, try to fallback to -mod=mod
+			// or not set if the go version < 1.14. Note we check this post-build in case the build brings
+			// the vendor directory up to date.
+			if !checkVendor() {
+				workspace.ModMode = project.ModMod
+				log.Println("The vendor directory is not consistent with the go.mod; not using vendored dependencies.")
+			}
+		}
 
-		pt := createPathTransformerFile(paths.newdir)
-		defer os.Remove(pt.Name())
+		if shouldInstallDependencies {
+			if workspace.ModMode == project.ModVendor {
+				log.Printf("Skipping dependency installation because a Go vendor directory was found.")
+			} else {
+				installDependencies(workspace)
+			}
+		}
 
-		writePathTransformerFile(pt, paths.realSrc, paths.root, paths.newdir)
-		setGopath(paths.root)
+		workspaces[i].Extracted = extract(workspace)
 	}
 
-	// check whether an explicit dependency installation command was provided
-	inst := util.Getenv("CODEQL_EXTRACTOR_GO_BUILD_COMMAND", "LGTM_INDEX_BUILD_COMMAND")
-	shouldInstallDependencies := false
-	if inst == "" {
-		shouldInstallDependencies = buildWithoutCustomCommands(buildInfo.ModMode)
-	} else {
-		buildWithCustomCommands(inst)
-	}
+	// Find all projects which could not be extracted successfully
+	var unsuccessfulProjects = []string{}
 
-	if buildInfo.ModMode == project.ModVendor {
-		// test if running `go` with -mod=vendor works, and if it doesn't, try to fallback to -mod=mod
-		// or not set if the go version < 1.14. Note we check this post-build in case the build brings
-		// the vendor directory up to date.
-		if !checkVendor() {
-			buildInfo.ModMode = project.ModMod
-			log.Println("The vendor directory is not consistent with the go.mod; not using vendored dependencies.")
+	for _, workspace := range workspaces {
+		if !workspace.Extracted {
+			unsuccessfulProjects = append(unsuccessfulProjects, workspace.BaseDir)
 		}
 	}
```

Comment:
Consider building this array at the same time as setting `workspaces[i].Extracted`, just to avoid having an extra for loop. I think this is the only use of `.Extracted`, so you could even get rid of that field if you do that.

Reply:
Sounds reasonable, I think this extra loop here is just the remains of a previous implementation. I will remove this in a follow-up PR if we think we can get this merged as is today.